### PR TITLE
Add validation plot after Bastos and O'Hagan (2009), along with example

### PR DESCRIPTION
**Pull request by prosku**
Adds a validation plot modeled after Fig. 1 in Bastos and O'Hagan (2009) to utils.py, along with an example on top of the Emulating_using_GPs.ipynb.

In the example, the validation looks strange for me in the plot of standardized errors against emulated values. As I am not familiar with the data set, maybe @duncanwp has an idea for whether that's reasonable?

No tests were added for this.

*References*
Bastos and O'Hagan (2009): Diagnostics for Gaussian Process Emulators, Technometrics, 51, 425-438. https://doi.org/10.1198/TECH.2009.08019

### CODE REVIEW

File: esem/utils.py [hunk 1]
```diff
@@ -91,6 +91,104 @@ def validation_plot(test_mean, pred_mean, pred_var, figsize=(7, 7), minx=None, m
...
+    sm.qqline(qq_plot.axes[2], line='45', fmt='k--')
+    axs[1, 0].set_xlim([end_pts[0][0], end_pts[1][0]])
+    axs[1, 0].set_ylim([end_pts[0][1], end_pts[1][1]])
+    axs[1, 0].set_ylabel('Standardized quantiles')
+
+    for i in range(0, np.shape(X_test)[1]):
+        # Slightly convoluted way to expand the parameters to match the shape of the outputs
+        expanded_params = np.broadcast_to(np.expand_dims(X_test.to_numpy()[:, i], axis=[i for i in range(1, len(errors_std.shape))]),
+                                          errors_std.shape)
+        axs[0, 1].scatter(expanded_params, errors_std, c=colors[i], label=X_test.columns[i], marker='.', alpha=alpha)
```
Comment (duncanwp) [inline esem/utils.py:148]: I've changed the label to use the column names but I realise this assumes X_test is a pd.DataFrame, we should probably check that first

File: esem/utils.py [hunk 2]
```diff
@@ -91,6 +91,104 @@ def validation_plot(test_mean, pred_mean, pred_var, figsize=(7, 7), minx=None, m
...
+    qq_plot = pp.qqplot(marker='.', markerfacecolor='k',
+                        markeredgecolor='k', alpha=alpha, ax=axs[1, 0])
+    end_pts = lzip(axs[1, 0].get_xlim(), axs[1, 0].get_ylim())
+    sm.qqline(qq_plot.axes[2], line='45', fmt='k--')
+    axs[1, 0].set_xlim([end_pts[0][0], end_pts[1][0]])
+    axs[1, 0].set_ylim([end_pts[0][1], end_pts[1][1]])
+    axs[1, 0].set_ylabel('Standardized quantiles')
+
+    for i in range(0, np.shape(X_test)[1]):
+        # Slightly convoluted way to expand the parameters to match the shape of the outputs
```
Comment (duncanwp) [inline esem/utils.py:145]: I've refactored this to tidy it up (and it should make it quicker too)

File: esem/utils.py [hunk 3]
```diff
@@ -91,6 +91,104 @@ def validation_plot(test_mean, pred_mean, pred_var, figsize=(7, 7), minx=None, m
...
+    # add hlines
+    axs[0, 1].axhline(y=-2, c=c_black, linestyle='--')
+    axs[0, 1].axhline(y=2, c=c_black, linestyle='--')
+    axs[0, 1].axhline(y=0, c=c_black, linestyle='--')
+    axs[0, 0].axhline(y=-2, c=c_black, linestyle='--')
+    axs[0, 0].axhline(y=2, c=c_black, linestyle='--')
+    axs[0, 0].axhline(y=0, c=c_black, linestyle='--')
+
+    ax = axs[1, 1]
+    # adapt validation_plot(...) from GCEm package:
```
Comment (duncanwp) [inline esem/utils.py:163]: I'd like all of the below to just use the `validation_plot` routine as much as possible. I realise it doesn't currently accept and axes, but it certainly could do. Or the plotting part could be factored out in to a seperate function that both plots can use

Comment (prosku) [reply esem/utils.py:163]: I agree, both options would be better than having so much redundant code

File: esem/utils.py (hunk 1 above)
Comment (prosku) [reply esem/utils.py:148]: Good idea to use more telling labels!
You mean check whether X_test was passed as a pd.DataFrame? But passing something else should also be allowed, right?

**Review by duncanwp — APPROVED** [review]
Looks good, thanks!

**Comment by duncanwp** [comment]
This is great, thanks!

**Pull request merged**